API 문서 만들기, Docker image 만들기

alex-gibeom opened the pull request:
안녕하세요 :)
벌써 코드숨의 마지막 주차네요.. 시간이 참 빠른 것 같습니다😂
멘토님들 덕분에 정말 많은 것을 배웠던 것 같습니다 🙇‍♂️🙇‍♂️

마지막 주차 잘 부탁드립니다! 😄

<br>

#### 이번 주차는 아래의 흐름대로 작업을 진행해볼까 합니다!
> 각 메인 작업 단위로 origin에서 PR을 받아 squash and merge를 통해 upstream으로 올리겠습니다

- [x]  프로젝트 세팅 (템플릿, 의존성)
- [x]  패키지 구조 분리 (헥사고날)
- [x]  Product 리팩토링
    - [x]  요청, 응답 DTO 분리
    - [x]  BDD 스타일로 테스트 코드 변경
    - [x]  예외 분리
    - [x]  JavaDoc 작성
- [x]  User 리팩토링
    - [x]  요청, 응답 DTO 분리
    - [x]  BDD 스타일로 테스트 코드 변경
    - [x]  예외 분리
    - [x]  JavaDoc 작성
- [x]  Session 리팩토링
    - [x]  요청, 응답 DTO 분리
    - [x]  BDD 스타일로 테스트 코드 변경
    - [x]  예외 분리
    - [x]  JavaDoc 작성
- [x]  RestDocs 적용
- [x]  Docker

alex-gibeom committed:
(#1) 프로젝트를 세팅한다

alex-gibeom committed:
(#2) Hexagonal Architecture를 적용한다

* refactor: 패키지 구조를 hexagonal-architecture 구조로 변경

* refactor: HelloController 삭제

ssisksl77 commented:
아앗! 혹시 금일은 기존에  작업하시던 프로젝트를 옮기기만 하신 걸까요?

alex-gibeom commented:
> 아앗! 혹시 금일은 기존에 작업하시던 프로젝트를 옮기기만 하신 걸까요?

넵 현재 리펙토링 중에 있습니다!
마지막 주차 repository에 8주간 배웠던 부분들을 모두 담아보고 싶어서, Restdocs 작업하기 전에 전체적으로 리팩토링을 진행하고있습니다 :)

내일 안으로 기존 코드 리팩토링을 마무리 할 예정입니다!
(최대한 리뷰해주시기 편하도록 커밋에 신경쓰겠습니다! 이상하거나 수상한 코드가 있다면 말씀해주세요 😃)

아래 PR 링크에 패키지 구조를 작성해보았는데 이상한 점이 없는지 한번만 확인해주시면 감사하겠습니다!
https://github.com/giibeom/codesoom-final-summary/pull/2

alex-gibeom committed:
(#3) Product 관련 프러덕션 및 테스트 코드를 모두 리팩토링한다

* feat(common): JsonUtil 생성 및 Product 객체로 테스트

* test(common): MockMvc 결과 값 인코딩 UTF-8로 설정

* refactor(product): ProductController 요청 DTO 분리

* feat(product): Mapstruct를 사용한 DTO to Entity 객체 매핑 기능

* refactor(product): ProductService에 요청 DTO 분리 적용

* test(common): Token, Id, 상품의 fixture 객체 생성

* refactor(product): ProductController 웹 테스트 BDD 스타일로 변경

* refactor(product): ProductService 유닛 테스트 BDD 스타일로 변경

* refactor(product): Product 유닛 테스트 BDD 스타일로 변경

* refactor(product): ProductService port 생성 (UseCase 인터페이스)

* feat(product): ProductController, Product 객체 내부의 public 메서드에 JavaDoc 작성

* refactor(product): ProductController 응답 DTO 분리

* feat(product): 공통 예외 처리에서 product 관련 예외 분리

* fix(auth): 응답 토큰 없을 때의 예외 테스트 깨지는 현상 수정
   - AccessDeniedException을 핸들러에서 묶어서 잡아버리면 인증 토큰이 없는 상황과 권한이 없는 상황을 구분할 수 없음
   - 방법을 찾기 전까지 AccessDenied 예외 핸들러는 주석처리

alex-gibeom committed:
(#4) User 관련 프러덕션 코드를 모두 리팩토링한다

* refactor(user): UserController 등록, 수정 RequestDto 분리 및 command(인터페이스) 생성

* refactor(user): UserController 등록, 수정 ResponseDto 분리 (Dto has a Entity 방식)

* refactor(user): User usecase 인터페이스(port)를 통해 Controller의 직접적인 의존 역전

* feat(user): Mapstruct를 사용한 DTO to Entity 객체 매핑 기능 추가 및 DozerMapper 삭제

* refactor(user): User 프러덕션 코드 리팩토링 및 JavaDoc 주석 작업

* refactor(user): 공통 예외 처리에서 user 관련 예외 분리

ssisksl77 commented:
> > 아앗! 혹시 금일은 기존에 작업하시던 프로젝트를 옮기기만 하신 걸까요?
> 
> 넵 현재 리펙토링 중에 있습니다! 마지막 주차 repository에 8주간 배웠던 부분들을 모두 담아보고 싶어서, Restdocs 작업하기 전에 전체적으로 리팩토링을 진행하고있습니다 :)
> 
> 내일 안으로 기존 코드 리팩토링을 마무리 할 예정입니다! (최대한 리뷰해주시기 편하도록 커밋에 신경쓰겠습니다! 이상하거나 수상한 코드가 있다면 말씀해주세요 😃)
> 
> 아래 PR 링크에 패키지 구조를 작성해보았는데 이상한 점이 없는지 한번만 확인해주시면 감사하겠습니다! [giibeom#2](https://github.com/giibeom/codesoom-final-summary/pull/2)

넵 아주 좋네요! 일단 코멘트를 작게 남겨보았습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/common/exception/ControllerErrorAdvice.java:
```diff
@@ -37,4 +38,53 @@ public ErrorResponse handleUserEmailIsAlreadyExisted() {
+    }
+
+    private String getViolatedMessage(ConstraintViolationException exception) {
+        String messageTemplate = null;
+        Set<ConstraintViolation<?>> violations = exception.getConstraintViolations();
+        for (ConstraintViolation<?> violation : violations) {
+            messageTemplate = violation.getMessageTemplate();
+        }
+        return messageTemplate;
```
조금 설명이 있으면 좋겠어요! for문을 돌면서 메시지를 덮는 것인가요?

alex-gibeom replied:
앗 확인해보니 7주차 베이스 코드에서 예외 핸들러들만 복붙해왔는데 그러다가 같이 껴들어간 코드 같습니다..!
PR 올리느라 아직 많이 검색해보지는 못했지만, 피드백 주신 해당 예외 핸들러는 Controller가 아닌 다른 곳에서 검증을 할 때 사용되는 `@Validated`에서 던지는 예외인 것 같습니다!
(더 자세한 내용은 PR 올린 후 계속 서칭해보겠습니다)

따라서 저의 프로젝트 구조 상 모든 데이터 유효성 검증은 Controller 매개변수의 `@Valid`를 통해서 이루어지고, 데이터 유효성 검증 실패 시 발생하는 예외는 `MethodArgumentNotValidException` 이므로 해당 핸들러에는 주석을 작성하였으며, 피드백 주신 부분의 코드는 사용되지 않는 것이 확인되어 삭제 처리 하였습니다!

#### [참고 자료]
- https://www.baeldung.com/spring-valid-vs-validated
- https://meetup.toast.com/posts/223

alex-gibeom committed:
(#5) User 관련 테스트 코드 및 프러덕션 코드를 모두 리팩토링한다

* refactor(user): User RequestDto,예외 핸들러 리팩토링

* refactor(user): UserController 웹 테스트 BDD 스타일로 변경, 인증 관련 given()절 분리

* fix(user): 회원 수정 로직에 비밀번호 변경 로직 추가

* refactor(user): UserService 유닛 테스트 BDD 스타일로 변경

* refactor(user): User 유닛 테스트 BDD 스타일로 변경 및 리팩토링

* refactor(user): User 관련 코드 매개변수 final 처리 및 주석 작업

alex-gibeom committed:
(#6) Session (로그인 인증) 관련 프러덕션 및 테스트 코드를 모두 리팩토링한다

* refactor(session): SessionController 웹 테스트 BDD 스타일로 변경 및 리팩토링

* refactor(session): AuthenticationService useCase 인터페이스(port)를 통해 객체 간의 직접적인 의존 역전

* refactor(session): AuthenticationService 유닛 테스트 BDD 스타일로 변경

* refactor(session): 공통 예외 처리에서 인증 관련 예외 분리

alex-gibeom commented:
기존 베이스 코드 대부분을 얼추 리팩토링이 끝난 것 같습니다
내일부터는 오늘 피드백 주시는 부분들 반영 후에, RestDocs 작업 시작해보려고 합니다!

3일 내내 열심히 진행해보았는데, 부족한 점이 보인다면 피드백 주시면 감사하게 학습하겠습니다 :)
혹시 코드 외적으로도 PR이나 커밋 방법 등의 보기 불편하신 점이 있다면 언제든 피드백 주시면 감사하겠습니다 😄😄

ssisksl77 commented:
뭔가 엄청난 것들이 바뀌었군요. 

뭔가 헥사고날 아키텍처에 대한 리뷰를 이어가고 싶으신 마음에 8주차 코드와 많이 다른 아키텍처를 연결하시고 계신 마음을 느꼈습니다.

헥사고날 아키텍처를 수행하면서 기존 MVC와 어떻게 다른 것으로 느껴졌는지 
어느부분이 좋았는지 어느부분이 별로인지 좀 더 코드레벨에서의 논의가 이어지면 어떨까 라는 생각을 했습니다.

ssisksl77 commented:
대부분 멋지게 해내셨고 어떤 부분은 갸우뚱스러운 부분도 있지만 이 부분들은 저절로 논의하면서 답을 찾아가지 않을까 싶습니다.

ssisksl77 commented:
[a866c4f](https://github.com/CodeSoom/spring-week8-assignment-1/pull/71/commits/a866c4f20cf7ccd3aff369a8087fb31f5d7225ab) 제목 왼쪽에 있는 커밋링크는 어떤 의도인가요?

ssisksl77 commented:
전체적인 아키텍처에 대한 설명보다는 
MVC와 비교하며 무엇을 어떻게 나누었는지 설명해주시면 감사하겠습니다.

giibeom commented:
> 헥사고날 아키텍처를 수행하면서 기존 MVC와 어떻게 다른 것으로 느껴졌는지 어느부분이 좋았는지 어느부분이 별로인지 좀 더 코드레벨에서의 논의가 이어지면 어떨까 라는 생각을 했습니다.

너무 좋습니다!
코드레벨에서 어느 부분이 개선해야될 곳인지 짚어주시면 더 빠르게 이해할 수 있을 것 같습니다 :)

alex-gibeom commented:
> [a866c4f](https://github.com/CodeSoom/spring-week8-assignment-1/pull/71/commits/a866c4f20cf7ccd3aff369a8087fb31f5d7225ab) 제목 왼쪽에 있는 커밋링크는 어떤 의도인가요?
`a866c4: (#5) User 관련 테스트 코드 및 프러덕션 코드를 모두 리팩토링한다.`

<br>

`커밋이 너무 많으면 리뷰하기 힘들다`는 피드백을 토대로 고민을 해보았고, 그 고민 결과로 `커밋 링크`를 통해 해당 **`작업의 총 정리 문서 scope`** 를 사용하고 관련 커밋들을 하나로 묶어버리면 어떨까라는 의도로 사용해보았습니다

제가 생각했었던 PR 방식의 흐름은 아래와 같습니다
- PR 하나는 작업에 대한 보고서 및 설계서이자, 요구 사항에 대한 Todo List 정리본이라고 생각하였습니다
- 현재 코드숨 리뷰의 방식이 하나의 PR을 merge 하기전에 리뷰를 하는 형식이 아닌 모든 작업을 하나의 PR로 통합하는 형태입니다 (많은 커밋이 정리되지 않고 쌓일 수 밖에 없는 형식)
- 따라서 **각 작업 단위**로 PR을 묶고, 한 PR의 관련 작업들을 `커밋 링크`에 모두 정리해놓았으며 squash and merge 를 통해 최종 커밋내역을 깔끔하게 정리해보았습니다
   - 각 작업 단위: 맨 위의 코멘트에서 작성한 메인 작업 단위
   - 예시) Product 리팩토링, User 리팩토링

<br>

#### 참고자료
https://github.com/CodeSoom/spring-week5-assignment-1/pull/80#issuecomment-1312718333
![image](https://user-images.githubusercontent.com/59248326/204987980-945ae59d-997a-441e-9621-aa1e88b58cbe.png)

alex-gibeom commented:
> 전체적인 아키텍처에 대한 설명보다는 MVC와 비교하며 무엇을 어떻게 나누었는지 설명해주시면 감사하겠습니다.

<br>

기존 MVC을 헥사고날로 변경하면서 나눈 것들과 느낀점을 아래에 정리해보겠습니다

1. useCase와 domain을 중심에 두고 나머지 외부적인 요소(controller, repostiroy)들은 모두 adapter라는 계층으로 집어넣습니다
   - MVC: DB를 향해 폭포수처럼 흘러가는 방식입니다
   - Hexagonal: UseCase와 domain을 중심으로 의존의 방향이 몰립니다

2. adapter 패키지의 in에는 요청이 들어오는 외부 환경에 관한 어댑터를, out에는 영속성 환경에 관한 어댑터를 집어 넣습니다
   - MVC: 계층 별(Controller/Service/Repository/Domain)로 별개의 패키지를 만들어 분리합니다
   - Hexagonal: application 계층을 중심으로 외부적인 요소는 adapter 패키지 하나로 묶고 각 어댑터는 port(인터페이스)를 통해 연결하여 application을 재사용할 수 있는게 장점이라고 생각합니다

<br>

아래 내용은 기존 코드를 변경하였지만, Hexagonal과는 별개로 MVC에서도 흔히 하는 방식을 따로 정리해보았습니다

1. interface(port)를 통해 Controller와 Service의 직접적인 의존을 역전시킵니다
2. 요청과 응답의 DTO를 서로 분리해줍니다.
3. 각 도메인 별로 관련 exception을 분리하여 책임을 분리해줍니다.

#### 전체적으로 느낀 관점 정리

- usecase 및 domain 로직은 다양한 환경에서 재사용할 수 있도록 구성합니다
- 어댑터는 외부 환경에 대한 요소를 usecase를 사용할 수 있도록 변환해주는 역할을 담당합니다
- 이로써 domain 로직의 변경은 오로지 `domain에 관한 요구 사항 변경`일 때만 수정이 일어나므로 SRP를 지킬 수 있는 것 같습니다
- 또한 모든 계층 간 port(인터페이스)를 통해 직접적인 의존을 역전시켜 독립적으로 분리하므로 추후 갈아끼우기가 정말 쉬울 것 같다는 생각도 들었습니다

헥사고날을 아직 자세하게 이해하진 못한 것 같습니다..! 정리한 관점 중에 제가 놓치고 있거나, 틀린 생각이 있다면 바로잡아주시면 감사히 습득하겠습니다 😊

<br>

---

> 헥사고날 아키텍처를 수행하면서 기존 MVC와 어떻게 다른 것으로 느껴졌는지 어느부분이 좋았는지 어느부분이 별로인지 좀 더 코드레벨에서의 논의가 이어지면 어떨까 라는 생각을 했습니다.

너무 좋습니다!
코드레벨에서 어느 부분이 개선해야될 곳(이상해보이는 곳)인지 짚어주시면 더 빠르게 이해할 수 있을 것 같습니다 :)

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/product/application/ProductService.java:
```diff
@@ -0,0 +1,56 @@
+        return productRepository.save(productCreateRequest.toEntity());
+    }
+
+    public Product updateProduct(final Long id,
+                                 final ProductUpdateRequest productUpdateRequest) {
+        Product product = findProduct(id);
+
+        product.update(productUpdateRequest.toEntity());
+
+        return product;
+    }
```
DTO에 대해서 궁금한 것이 있습니다!

헥사고날 아키텍처에서는 유스 케이스에 관한 재사용성을 중요하게 생각한다고 느꼈는데, 그렇다면 application으로 들어오기 전에 어댑터(Controller)에서 toEntity()를 통해 엔티티를 변환한 후 넘겨주는 것이 좋을까요?
아니면 그냥 application 상단에서 변환작업을 해주어도 상관 없는 걸까요?

저는 일단 service 내부에서 변환 작업을 진행한 이유는 entity를 controller에서 매개변수로 넘겨준다면, 어느 개발자의 혼동에 의해 37번째 줄에서 영속성을 통해 가져온 DB를 사용해야되는데 실수로 매개변수를 사용하는 참사를 일으킬 가능성이 있고, 그렇게 되면 null을 사용할 수 있다는 가능성이 있기에 혼동을 방지하고자 service에서 entity로 변경하는 방식을 선택하였었습니다!

ssisksl77 replied:
좋습니다.

1. 
일단 Dto라고 생각해보지 말고 updateProduct에 필요한 매개변수를 객체로 만들어놓은 것이라고 생각해보죠!

`updateProduct` 는 `id` 와 `productUpdateRequest` 가 있습니다.

`productUpdateRequest` 라는 이름만 보아서는 상품갱신요청이기 때문에 바꾸고 싶은  `id`도 함께 들어있어야 하는 느낌이 듭니다.

2. 이 관점에서 다시 보면 DTO는 두개의 레이어를 연결해주는 다리 역할을 합니다.  Entity가 전파되지 않겠다는 의지의 표현은 아주 훌륭합니다.  중요한 것은 이것입니다. 역시 제대로 하려면 객체가 너무  많아집니다. 하지만 그만큼 객체가 하는 일이  적어지기 때문에 장황해 보이긴 하겠지만 수정은 그만큼 쉽게될 것입니다.

ssisksl77 replied:
그러므로

Entity를 직접 넘기는 것은 다른시각으로 보면 Repository 레이어가 노출되는 것이기도 합니다. (단순 데이터가 아니라 여러 기능이 종합적으로 존재하는)

그렇기에 실수를 할 수 있는 여지가 있다고 판단하는 것은 합리적입니다.

> * 어댑터는 외부 환경에 대한 요소를 usecase를 사용할 수 있도록 변환해주는 역할을 담당합니다

헥사고날의 특징  중 하나는 어댑터를 갈아 끼우기 쉬워야 한다는 점입니다. 가령 헥사고날에서 현재 REST API 를 만들고 있지만 콘솔 앱을 만들기도  쉽도록 잘 갈아끼워져야합니다. 

현재 구조는 잘 갈아끼워질까요?
그렇다/아니다 에 대한 본인의 생각과 그에 대한 근거를 코드레벨로 한번 공유해주세요!

alex-gibeom replied:
먼저 HTTP 환경 말고는 경험해보지 못해서 잘 감이 오지는 않는 점 양해 부탁드립니다..!

```java
public Product updateProduct(final Long id,
                             final ProductUpdateRequest productUpdateRequest) {
    Product product = findProduct(id);
    product.update(productUpdateRequest.toEntity());
    return product;
}
```
제가 생각하기에는 현재 두번째 매개변수 `ProductUpdateRequest`는 **인터페이스**로 구성되어있고 이를 구현하는 `ProductUpdateRequestDto`가 현재 어댑터(ProductController)에서 요청을 받는 구조로 되어있습니다
따라서 **직접적으로 객체가 의존되어있지 않기 때문**에 외부 어댑터가 콘솔앱에서 들어올 경우
새로운 ProductUpdateRequest의 구현체(ex.`ProductUpdateConsoleRequestDto`)를 만들어서 콘솔앱의 데이터 형태를 어댑터에서 변환하여 유스케이스로 넣어준다면 유스케이스에서의 수정 없이 갈아끼울 수 있을 것으로 생각이 듭니다!
제가 놓치고 있는 관점이 있을까요?

ssisksl77 replied:
아주 잘 설명하셨습니다.

여기서 말씀하신대로 controller의 구현만 바뀌면 HTTP가 아니더라도 가능하게  될 것입니다.

ssisksl77 replied:
하지만 다시 생각해보면 현재 만약에 console 앱을 위한 모듈을 추가로 넣겠다 하면
어디에 넣어야 할지 고민이 될 것입니다.

ssisksl77 replied:
아마 in/console 같은 패키지가 될 것 같습니다만

ssisksl77 replied:
헥삭고날 아키텍처를 좀 더 파보고 싶다면 아마 제대로 모듈화를 하셔야할 것입니다. 
그레이들의 멀티모듈을 사용해보심이 그 시작일 것입니다.

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/product/domain/Product.java:
```diff
@@ -0,0 +1,75 @@
+        this.imageUrl = imageUrl;
+    }
+
+    /**
+     * 상품 정보를 수정합니다. <br>
+     * 필드가 null일 경우 수정하지 않습니다.
+     *
+     * @param updateProduct 수정할 상품 정보
+     */
+    public final void update(Product updateProduct) {
+        updateName(updateProduct.getName());
+        updateMaker(updateProduct.getMaker());
+        updatePrice(updateProduct.getPrice());
+        updateImageUrl(updateProduct.getImageUrl());
+    }
```
JavaDoc 주석은 이런식으로 작성하면 되는 것일까요?
전체적으로 이런 느낌으로 작성하였는데 저한테만 잘 읽히는 방식일지 걱정됩니다

ssisksl77 replied:
저는 아주 좋아보입니다. 좀 더 자세히보니 `필드가 null일 경우 수정하지 않습니다.`  를 좀 더 풀어서 쓰셔도 좋을 것 같습니다.

alex-gibeom replied:
전체적으로 한번 더 다듬으면서 너무 간략한 설명들은 좀 풀어서 작성해야겠네요!

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/user/application/UserService.java:
```diff
@@ -0,0 +1,74 @@
+        return user;
+    }
+
+    public User updateUser(final Long id,
+                           final UserUpdateRequest userUpdateRequest,
+                           final Long userId) throws AccessDeniedException {
+        // TODO: 인가 로직 AOP로 분리
+        if (!id.equals(userId)) {
+            throw new AccessDeniedException("Access denied");
+        }
+
+        User user = findUser(id);
+
+        user.update(userUpdateRequest.toEntity(), passwordEncoder);
+
+        return user;
+    }
```
강의 베이스 코드에서는 53번째 줄에서 요청 데이터의 pathVariable로 올라온 id와 토큰의 id가 일치하는지 `인가를 하는 로직`은 `updateUser` 메서드에 있는 것이 좀 신경쓰입니다
SRP에 위배된다고 생각해서 인터셉터나 AOP를 통해 분리하는 것이 좋아보인다고 생각이 드는데 어떠신것 같나요?

ssisksl77 replied:
좋은 생각입니다.

`인터셉터 -> 컨트롤러`  로직 범위 어딘가에 있으면 좋겠다는 생각이 드는군요!

alex-gibeom replied:
해당 Service의 updateUser() 메서드를 포인트컷으로 잡아 AOP의 `@Before`를 통해 인가 로직을 분리해보겠습니다 :)

giibeom commented on app/src/test/java/com/codesoom/assignment/presentation/ProductControllerTest.java:
```diff
@@ -0,0 +1,596 @@
+    }
+
+
+    @Nested
+    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+    class list_메서드는 {
+
+        @Test
+        @DisplayName("200 코드로 응답한다")
+        void it_responses_200() throws Exception {
+            ResultActions perform = mockMvc.perform(
+                    get(REQUEST_PRODUCT_URL)
+            );
+
+            perform.andExpect(status().isOk());
+
+            verify(productService).getProducts();
+        }
+    }
```
restDocs 작업중에 발생한 오류에 관해 질문이 있습니다..!
현재 상황은 해당 클래스 맨 상단에 `@AutoConfigureRestDocs`를 선언하고, 107번째 줄 코드 (`perform.andExpect(status().isOk());`)에 추가로 `.andDo(document("get-products"))`를 붙여놓고 테스트를 실행하였습니다.
그랬더니 아래와 같이 `NullPointerException`이 발생합니다

<img src="https://user-images.githubusercontent.com/59248326/205047248-d9f283ce-eb49-4198-a7c7-eda5cca56744.png" width="100%" />


서칭해보니 `@Nested`문제인 것 같은데 해당 클래스의 테스트 어노테이션을 `@SpringBootTest`로 변경하고 아래와 같이 Nested 테스트 클래스를 설정해주었더니 동작하긴 합니다만...
```java
@SpringBootTest // 스프링 테스트로 변경
@AutoConfigureMockMvc
@AutoConfigureRestDocs
@DisplayName("ProductController 웹 유닛 테스트")
class ProductControllerTest {
    private static final String REQUEST_PRODUCT_URL = "/products";

    @Autowired
    private MockMvc mockMvc;

    @Nested
    @SpringBootTest // Nested에서 한번 더 선언
    @AutoConfigureMockMvc
    @AutoConfigureRestDocs
    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
    class list_메서드는 {

        @Autowired
        private MockMvc mockMvc;

        @Test
        @DisplayName("200 코드로 응답한다")
        void it_responses_200() throws Exception {
            ResultActions perform = mockMvc.perform(
                    get(REQUEST_PRODUCT_URL)
            );

            perform.andExpect(status().isOk());
            perform.andDo(document("get-products"));

            verify(productService).getProducts();
        }
    }
}
```


위처럼 모든 테스트에 같은 설정값을 선언해주는 방식으로 진행하는 것이 맞는걸까요? 아니면 실 객체 테스트를 만들겸 RestDocs를 사용하기 위해 따로 `@Nested` 없는 테스트를 만들어서 그쪽에다가 작업을 하는 것이 더 나을까요?
방법을 잘 모르겠습니다....😂😂
위의 예시처럼 하게 된다면 application context가 선언된 곳마다 계속 새로 생성됩니다

참고자료 : https://github.com/spring-projects/spring-restdocs/issues/490

giibeom replied:
`@AutoConfigureMockMvc`를 사용하지 않고 MockMvc를 설정하는 방법을 한번 찾아보려고 합니다

alex-gibeom committed:
(#7) RestDocs 세팅 및 Product API를 문서화 한다

* chore: RestDocs 문서 관련 gradle 설정

* feat: RestDocs 문서 생성을 위한 provider, util 생성

* docs: Product API asciidoc 스니펫 생성 작업

* docs: Product API 문서화

alex-gibeom commented on app/src/test/java/com/codesoom/assignment/presentation/RestDocsMockMvcProvider.java:
```diff
@@ -0,0 +1,49 @@
+import static org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.documentationConfiguration;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+
+/**
+ * MockMvc을 설정하여 제공하는 객체입니다. <br>
+ * 해당 객체를 상속 받을 경우 `@WebMvcTest` 혹은 `@SpringBootTest`를 통해 WebApplicationContext를 제공해야됩니다.
+ */
+@Import(MockMvcCharacterEncodingCustomizer.class)
+@ExtendWith({RestDocumentationExtension.class})
+public class RestDocsMockMvcProvider {
+
+    protected MockMvc mockMvc;
+
+    @Autowired
+    private WebApplicationContext context;
+
+    @Autowired
+    private Filter springSecurityFilterChain;
+
+    @BeforeEach
+    void setUpMockMvc(final RestDocumentationContextProvider restDcosContextProvider) {
+        this.mockMvc = MockMvcBuilders.webAppContextSetup(context)
+                .apply(documentationConfiguration(restDcosContextProvider))
+                // Spring Security 필터 체인을 주입합니다.
+                .addFilter(springSecurityFilterChain)
+                // 테스트의 결과 값을 UTF-8로 설정해줍니다.
+                // (Spring Boot 2.2부터는 Spring에서 charset을 사용하지 않고 기본 인코딩 문자도 더 이상 UTF-8이 아닙니다.)
+                // Ref: https://github.com/spring-projects/spring-framework/issues/22788
+                .alwaysDo(result -> result.getResponse().setCharacterEncoding(StandardCharsets.UTF_8.name()))
+                .alwaysDo(print())
+                .build();
+    }
+}
```
현재 Spring Boot 2.3.x 버전에서 `@Nested`를 활용한 테스트를 진행할 때 이너 클래스와 이를 감싸고 있는 클래스가 서로 다른 Application Context를 사용하여 config 값들이 공유 되지 않는 이슈가 있습니다
따라서 `@AutoConfigureRestDocs`를 맨 상단에 설정하면 이너 클래스들이 주입을 받지 못해 테스트를 정상적으로 실행하지 못했습니다
이를 해결하기 위해 MockMvc와 RestDocs의 설정을 `@AutoConfigureMockMvc`와 `@AutoConfigureRestDocs`를 통해 자동 주입을 하는 방식이 아닌,  위와 같이 직접 설정해주는 방식을 사용했는데, 여기에서 궁금한 점이 있습니다!
`RestDocsMockMvcProvider` 클래스를 분리하고 실제 사용하는 웹 테스트에서 확장하여 `mockMvc`를 사용하는 방식으로 설계해보았는데 좀 더 좋은 방법이 있을지 궁금합니다!

#### [참고자료]
- [https://github.com/spring-projects/spring-restdocs/issues/490#issuecomment-373333848](https://github.com/spring-projects/spring-restdocs/issues/490#issuecomment-373333848)
- [https://github.com/spring-projects/spring-framework/issues/19930](https://github.com/spring-projects/spring-framework/issues/19930)

ssisksl77 replied:
좋습니다. 재사용하는 컨텍스트를 확장으로 푸는 것을 스프링에서 테스트를 할 때 일반적인 방식입니다.

alex-gibeom commented:
<details>
<summary>Resolve conversation</summary>
<div markdown="1">

아래의 링크의 **참고 자료**에 Product API만 RestDocs를 구성해보았는데, RestDocs의 사용경험이 없다보니 Best Practice를 잘 모르겠어서 어떤식으로 구성해야될지 잘 모르겠습니다
보완점을 말씀해주시면 감사히 적용해보겠습니다!
https://github.com/giibeom/codesoom-final-summary/pull/7#issue-1472706775

</div>
</details>

alex-gibeom commented:
멘토님 한가지 질문이 있습니다
현재 IntelliJ에서 `./gradlew assemble`을 통해 생성한 jar 파일(`{프로젝트 위치}/app/build/libs/app.jar`)이 들어있는 `lib` 폴더를 도커를 실행할 때 `/home/api-server`라는 폴더에다 마운트시키는 명령어를 아래와 같이 입력하였습니다
하지만 아래 사진처럼 새로운 폴더(`libs;D`)가 애플리케이션 내부에 생성되는데 도커를 통해 들어간 리눅스에서는 app.jar가  보이지 않아 마운트도 정상적으로 되지 않은 것 같은데 너무 어려운 것 같습니다... (제 환경은 window 10에서 gitbash를 사용하고 있습니다..!)
어떤식으로 관련 오류를 검색, 해결해야할까요?

#### 제가 실행한 도커 실행 명령어 입니다 (강의와 동일한 명령어)
`docker run -it --name api-server -v $(pwd)/app/build/libs:/home/api-server openjdk:11 bash`


#### 현재 상황

<img src ="https://user-images.githubusercontent.com/59248326/205443261-f5a632a2-c239-4237-8f21-54f7900fe625.png" width = "70%" />

#### 강의 상황

<img src ="https://user-images.githubusercontent.com/59248326/205444008-51c7e36d-1b90-48d8-9eac-3c3be495f589.png" width = "70%" />



<br>

비슷한 상황은 겪은 사람이 stackoverflow에 글을 올려놓은 것 같은데 답변이 달리지 않았네요..!
- https://stackoverflow.com/questions/65861099/docker-mount-on-windows-strage-characters-on-the-host-path

alex-gibeom commented:
> 대부분 멋지게 해내셨고 어떤 부분은 갸우뚱스러운 부분도 있지만 이 부분들은 저절로 논의하면서 답을 찾아가지 않을까 싶습니다.

추가로 궁금한 점이 있습니다!
이번 주 초반에 리뷰를 해주신부분중에 "갸우뚱 스러운 부분"이 어느어느 곳인지도 궁금합니다!

ssisksl77 commented:
> 멘토님 한가지 질문이 있습니다 현재 IntelliJ에서 `./gradlew assemble`을 통해 생성한 jar 파일(`{프로젝트 위치}/app/build/libs/app.jar`)이 들어있는 `lib` 폴더를 도커를 실행할 때 `/home/api-server`라는 폴더에다 마운트시키는 명령어를 아래와 같이 입력하였습니다 하지만 아래 사진처럼 새로운 폴더(`libs;D`)가 애플리케이션 내부에 생성되는데 도커를 통해 들어간 리눅스에서는 app.jar가 보이지 않아 마운트도 정상적으로 되지 않은 것 같은데 너무 어려운 것 같습니다... (제 환경은 window 10에서 gitbash를 사용하고 있습니다..!) 어떤식으로 관련 오류를 검색, 해결해야할까요?
> 
> #### 제가 실행한 도커 실행 명령어 입니다 (강의와 동일한 명령어)
> `docker run -it --name api-server -v $(pwd)/app/build/libs:/home/api-server openjdk:11 bash`
> 
> #### 현재 상황
> <img alt="" width="70%" src="https://user-images.githubusercontent.com/59248326/205443261-f5a632a2-c239-4237-8f21-54f7900fe625.png">
> 
> #### 강의 상황
> <img alt="" width="70%" src="https://user-images.githubusercontent.com/59248326/205444008-51c7e36d-1b90-48d8-9eac-3c3be495f589.png">
> 
> 
> 비슷한 상황은 겪은 사람이 stackoverflow에 글을 올려놓은 것 같은데 답변이 달리지 않았네요..!
> 
> * https://stackoverflow.com/questions/65861099/docker-mount-on-windows-strage-characters-on-the-host-path

안녕하세요.

답변이 늦었습니다.

해당 문제를 확인하기위해 여러가지 방법으로 문제를 재연해보려고 했지만 재연이 되지 않고있습니다. 제 테스트 환경은 macos 입니다.

ssisksl77 commented:
application과 domain의 차이가 applicatino은 비즈니스로직이 들어있는 것이고, domain 로직은 데이터베이스 관련된 코드로 보입니다.

이것이 맞다면 패키지명은 조금 수정이 필요하겠습니다. application -> domain 으로 변경되고 domain 은 아마 repository나 entity 같은 것이 되어야겠습니다.

ssisksl77 commented:
헥사고널은 또한 비즈니스로직과 다른 플러그인(ex. Controller) 와는 물리적 독립이 된 것이 필요합니다.
그래야 갈아끼는 것이 용이합니다.

즉, `product` 패키지 안에 `adapter` 가 함께 있는 것은 독이될 수 있습니다. 왜냐하면 어댑터를 갈아끼우기 쉽지 않을 테니까요.

ssisksl77 commented:
제가 헥사고널 아키텍처로 아주 많이 프로젝트를 진행한 것은 아니지만 `application` 은 대게 빌드단위를 말합니다. 즉, 하나의 웹어플리케이션이 빌드되는 단위를 말하므로, application 은 여러 adpater 들을 선택하는 역할을 합니다. 

좀 더 구체적으로 말한다면, REST API 컨트롤러를 붙인 `application` app.jar 를 만들 것인지, GraphQL 컨트롤러를 붙인 `application` app.jar 를 만들지말입니다.

alex-gibeom commented:
> 해당 문제를 확인하기위해 여러가지 방법으로 문제를 재연해보려고 했지만 재연이 되지 않고있습니다. 제 테스트 환경은 macos 입니다.

도와주시려고 하신 점 감사드립니다
해당 부분은 Windows의 문제인 것으로 확인되어 폴더명을 아래와 같이 변경하였더니 해결되었습니다
`docker run -it --rm --name api-server -v "c:\\codesoom\\codesoom-final-summary\\app\\build\\libs":/home/api-server -p 80:8080 openjdk:11 bash -c "java -jar /home/api-server/app.jar"`

윈도우는 폴더의 path를 `/`가 아닌 `\\`으로 해야되나보네요...
얼른 mac으로 갈아타야겠습니다 😂

alex-gibeom commented:
> 헥사고널은 또한 비즈니스로직과 다른 플러그인(ex. Controller) 와는 물리적 독립이 된 것이 필요합니다. 그래야 갈아끼는 것이 용이합니다.
> 
> 즉, `product` 패키지 안에 `adapter` 가 함께 있는 것은 독이될 수 있습니다. 왜냐하면 어댑터를 갈아끼우기 쉽지 않을 테니까요.

아하 어댑터는 갈아끼우는 부품으로 생각했을 때는 비지니스 로직과 합쳐져있으면 좋지 않겠네요
여기서 adapter 안에 Product 관련 컨트롤러, User 관련 컨트롤러, Session 관련 컨트롤러와 r관련 요청 응답 DTO를 모두 모아놓으면 관리하기 힘들거 같다는 생각이 듭니다
그렇다면 adapter안에다가 관련 도메인별로 패키지를 만들어 정리해놓는게 좋을까요?

#### 예시: (dc43f2d)

<img src="https://user-images.githubusercontent.com/59248326/205490606-54998af4-8023-4190-9cea-f89febc39a79.png" width="50%" />

alex-gibeom commented:
> 제가 헥사고널 아키텍처로 아주 많이 프로젝트를 진행한 것은 아니지만 `application` 은 대게 빌드단위를 말합니다. 즉, 하나의 웹어플리케이션이 빌드되는 단위를 말하므로, application 은 여러 adpater 들을 선택하는 역할을 합니다.
> 
> 좀 더 구체적으로 말한다면, REST API 컨트롤러를 붙인 `application` app.jar 를 만들 것인지, GraphQL 컨트롤러를 붙인 `application` app.jar 를 만들지말입니다.

이 부분에 대해서는 코드숨 과정이 끝나고 추가로 공부를 진행하며 학습해보겠습니다!
새로운 관점을 열어주셔서 정말 감사드립니다

alex-gibeom committed:
(#8) adapter 패키지를 외부로 분리한다

ssisksl77 commented:
> > 헥사고널은 또한 비즈니스로직과 다른 플러그인(ex. Controller) 와는 물리적 독립이 된 것이 필요합니다. 그래야 갈아끼는 것이 용이합니다.
> > 즉, `product` 패키지 안에 `adapter` 가 함께 있는 것은 독이될 수 있습니다. 왜냐하면 어댑터를 갈아끼우기 쉽지 않을 테니까요.
> 
> 아하 어댑터는 갈아끼우는 부품으로 생각했을 때는 비지니스 로직과 합쳐져있으면 좋지 않겠네요 여기서 adapter 안에 Product 관련 컨트롤러, User 관련 컨트롤러, Session 관련 컨트롤러와 r관련 요청 응답 DTO를 모두 모아놓으면 관리하기 힘들거 같다는 생각이 듭니다 그렇다면 adapter안에다가 관련 도메인별로 패키지를 만들어 정리해놓는게 좋을까요?
> 
> #### 예시: ([dc43f2d](https://github.com/CodeSoom/spring-week8-assignment-1/commit/dc43f2d4f364fca41248ea64c6f14a417ef8b225))
> <img alt="" width="50%" src="https://user-images.githubusercontent.com/59248326/205490606-54998af4-8023-4190-9cea-f89febc39a79.png">

아키텍처 라는 것은 절충의 연속일 것입니다. 지금 구조로도 좋습니다. 어댑터를 갈아끼울일이 적다면 도메인별로 크게 나누는 것이 더 도움될 수 있습니다. 항상 어떤 구조가 좋다기보다 상황에 맞게 구조를 변화할 것입니다.

아마도 어떤 상황에서는 뭐가좋다 그런 공식을 알기보다는 고민하는 능력을 키우는 것이 도움되실 것입니다.

오늘의 답이 내일에는 틀릴 것이니까요

alex-gibeom committed:
(#9) 패키지 구조를 수정하고, 리팩토링을 진행한다

* refactor(product): 패키지 구조 변경 및 리팩토링

- 멤버, 매개변수 final 작업
- Transactional readonly 작업

* refactor(auth): 패키지 구조 변경 및 리팩토링

- 멤버, 매개변수 final 작업
- claim name static 멤버로 추출

* refactor(user): 패키지 구조 변경 및 1차 리팩토링

- 멤버, 매개변수 final 작업

* refactor(role): 패키지 구조 변경 및 1차 리팩토링

- 멤버, 매개변수 final 작업

alex-gibeom committed:
(#10) 회원 등록, 수정 비지니스 로직을 리팩토링한다

* refactor(user): 회원 등록, 수정 리팩토링

- toEntity()에서 비밀번호 인코딩 작업 진행
- MapStruct 매퍼 객체 삭제

* test: 패키지 명 수정

- domain -> application
- repository -> domain

* refactor(role): 유저의 권한 리팩토링

* refactor(user): 유저 수정 비지니스 로직 중 인가 로직 AOP로 분리

alex-gibeom committed:
(#11) 모든 도메인 엔티티의 기본키 생성을 데이터베이스로 위임한다

alex-gibeom committed:
docs: 코드숨 과정 총 정리 및 모든 주차 리뷰 링크 정리
